Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,128 @@
+complete sup-semilattice. (In literature, both variants occur.)
+
+* `is_integral`: a quantale for which `1 = ⊤`
+* `is_comm`: a commutative quantale, satisfying `x * y = y * x`
+* `is_idem`: an idempotent quantale, satisfying `x * x = x`
+
+## Naming conventions
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the left-residuation of `y` over `x`;
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the right-residuation of `y` over `x`;
+
+## References
+
+[Topology via Logic][vickers1989]
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale> (Categorical definition - not followed here)
+
+## TODO
+
+Additive quantales and multiplicative quantales both occur in literature.
+For sequences of actions, usually a multiplicative quantale is used, but for
+describing timing of actions (for example) the additive quantale of max-plus algebra
+a.k.a. tropical algebra, can be used. Therefore, extending the definitions with
+addition is - I think - relevant future work.
+
+-/
+
+/-- A quantale is a monoid distributing over a complete sup-semilattice.
+
+    Technically, I would prefer to define it as
+    `class Quantale (α : Type*) extends Monoid α, CompleteSemilatticeSup α where ...`
+    however, if I do that currently, the elements ⊤ and ⊥ are not automatically defined.
+    I could do that separately, but it seems more sensible to push the definition
+    of those to CompleteSemmilatticeSup in the library, as they are already making
+    sense there and having them should not break anything (in theory).
+
+    The reason not to consider a quantale as a complete lattice, is because the
+    morphisms between quantales usually only preserve the join and not the inf.
+    In fact, when seen as a generalization of locales, the monoid multiplication
+    usually takes the place of the inf.
+-/
+class Quantale (α : Type*) extends Monoid α, CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+section
+
+variable (α : Type _)
+variable [Quantale α]
+
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
+
+namespace Quantale
+
+/- Basic definitions on quantales -/
+
+/-- A quantale is integral if its unit and top element coincide -/
+def is_integral [Quantale α] := (⊤ : α) = 1
+
+/-- A quantale is commutative if its monoid satisfies `x * y = y * x` -/
+def is_comm [Quantale α] := ∀ x y : α , x * y = y * x
+
+/-- A quantale is idempotent if its monoid satisfies `x * x = x` -/
+def is_idem [Quantale α] := ∀ x : α, x * x = x
```

Comment:
These should be `IsIntegral`, `IsComm` and `IsIdem`. And presumably you would want them to be `class`es, right? Although, I'm not yet convinced that you want them to be mix-in classes.

For example, you want to conclude that a commutative quantale is a commutative monoid, no? What I'm arguing for here would be `CommQuantale (α : Type*) extends CommMonoid α, CompleteLattice α where ...`

Reply:
I'll change the spelling, thanks.
Not sure if they need to be classes... what would be the arguments for and against?
And what is a mix-in class?

Reply:
Well, if `IsComm` were not a class, then every time you wanted to use the fact about `CommMonoid`s for your commutative quantale, you would need to first activate `Quantale` + `IsComm` → `CommMonoid` instance manually. This would be a pain, and it's probably not what you want. For the others, I'm less sure whether they should be classes, as I don't really know what the use cases are.

Now, as for a mix-in, this is what these classes would be if you just changed them to `class` instead of `def` right now (and, if you did that, you should explicitly mark their type as `Prop`). A mix-in is a class which takes another class as an argument, but only adds `Prop`-valued fields. An example of this in the library would be `IsDomain`. In contrast, we don't use this pattern everywhere. For example, we don't write `Ring R` with `IsComm R` to specify a commutative ring. Instead, we write `CommRing R`, and the structure `CommRing` instead `extends` the structure `Ring`. This means that a `CommRing` carries with it a ring structure, instead of needing one supplied to it.

Reasons for preferring one to the other are related to convenience, performance or parallelism, or yet other reasons. In this case, I would almost certainly argue that `IsComm` should *not* be a mix-in, if for no other reason than we want the parallelism with the rest of the library. You can read more about the algebraic hierarchy in Mathlib in the book "Mathematics in Lean".

Reply:
My guess is that all properties are "at the same level as Comm" in this case, so I've updated the file to reflect this. Of course, we will only know once we start to actually prove things. I now have based the file mostly on the wikipedia entry on Quantales, which means there is one theorem that "obviously" needs to be proven, namely that every IdemIntegralQuantale is a Order.Frame. Will work on that shortly - but I was hoping the definitions can be approved before that?